tests/test_helper/bats/ovsdb_schema_upgrade.bats

```diff
@@ -61,20 +61,47 @@ setup() {
     local upgrade_container_index=0
     local old_container_index=0
     local containers_size="${#containers_to_upgrade[@]}"
+    local upgraded_size=0
 
     for ((container_index = 0; container_index < "$containers_size"; container_index++)) do
         local container="${containers_to_upgrade[$container_index]}"
         # Upgrade MicroOVN cluster, one cluster member at a time.
         install_microovn "$MICROOVN_SNAP_PATH" "$container"
+        upgraded_size=$(( "$upgraded_size" + 1 ))
+
+        # Check if the node is fully online. If it's not, then we have a dqlite schema upgrade and need to upgrade all nodes first.
+        # If all nodes are reporting ONLINE, then we can move on to check the ovsdb schema before upgrading all nodes.
+        wait_ovn_services "$container"
+        run wait_microovn_online "$container" 10
+        if [ "$status" = 0 ] ; then
+          echo "# ($container) reports all systems are reachable. No dqlite schema update detected"
+          break
+        fi
+
+        # If we upgraded all systems because of a dqlite schema update, ensure all nodes are online.
+        if [ "$upgraded_size" -eq "$containers_size" ] ; then
+          grep -q 0 <<< "$status"
+        fi
+    done
+
+    for ((container_index = 0; container_index < "$containers_size"; container_index++)) do
+        # If the current container index is larger or equal to the upgraded size, then we haven't run the update for that node.
+        if [ "$container_index" -ge  "$upgraded_size" ]; then
+            local container="${containers_to_upgrade[$container_index]}"
+            # Upgrade MicroOVN cluster, one cluster member at a time.
+            echo "# Upgrading ($container) before ovsdb schema verification"
+            install_microovn "$MICROOVN_SNAP_PATH" "$container"
+            upgraded_size=$(( "$upgraded_size" + 1 ))
+
+            wait_ovn_services "$container"
+            wait_microovn_online "$container" 30
+        fi
 
         # break out of the loop if every container has been upgraded
-        if [ $(( "$container_index" + 1 )) -eq  "$containers_size" ]; then
-            break
+        if [ $(( "$container_index" + 1 )) -eq  "$containers_size" ] || [ "$upgraded_size" -eq "$containers_size" ]; then
+          break
```

Comment:
If I'm reading this correctly, we will break out of the "main" loop if the previous "inner" loop, that you added, already  upgraded all the hosts, correct? That makes sense as we can't perform the upgrade checks via API if the dqlite schema is changing. However there's this block of code at the end of this main loop
```
# check that the schema upgrade was not triggered yet.
        local current_sb=""
        local current_nb=""
        current_nb=$(get_current_ovsdb_schema_version "$probe_container" "nb")
        current_sb=$(get_current_ovsdb_schema_version "$probe_container" "sb")
        assert [ "$current_nb" == "$original_nb" ]
        assert [ "$current_sb" == "$original_sb" ]
```
that verifies that the OVN schema change was not triggered until every node is upgraded. These function do not rely on the microovn API and therefore they could (and should) be executed after every host upgrade even if the dqlite schema is changing.

nitpick: your additions in this file have inconsistent indentation (2 vs 4 spaces).

Reply:
Good point, I've moved that check up into the first loop as well, and it will run after each system upgrades, except for the last one. On that one, we ensure that all nodes are ONLINE, and then check that the ovsdb schema update has been completed.